src/prepared_statement.c

```diff
@@ -0,0 +1,387 @@
+#include "bouncer.h"
+
+#include <usual/crypto/csrandom.h>
+#include <usual/hashtab-impl.h>
+
+static PgPreparedStatement* prepared_statements;
+queryid_t next_unique_query_id;
+
+/*
+ * Like uthash HASH_DELETE, but doesn't remove item from hash table just
+ * unlink element from the doubly-linked-list.
+ */
+#define HASH_UNLINK(hh, head, delptr)									\
+	do {																\
+		struct UT_hash_handle *_hd_hh_del = &(delptr)->hh;				\
+		if (_hd_hh_del->prev != NULL) {									\
+			HH_FROM_ELMT((head)->hh.tbl, _hd_hh_del->prev)->next = _hd_hh_del->next; \
+		} else {														\
+			DECLTYPE_ASSIGN(head, _hd_hh_del->next);					\
+		}																\
+		if (_hd_hh_del->next != NULL) {									\
+			HH_FROM_ELMT((head)->hh.tbl, _hd_hh_del->next)->prev = _hd_hh_del->prev; \
+		} else {														\
+			_hd_hh_del->tbl->tail = HH_FROM_ELMT(_hd_hh_del->tbl, _hd_hh_del->prev); \
+		}																\
+	} while (0)
+
+static PgPreparedStatement*
+create_prepared_statement(PgParsePacket *pkt)
+{
+	size_t query_len = strlen(pkt->query) + 1;
+	size_t num_parameters = pkt->num_parameters * 4;
+	PgPreparedStatement *ps = (PgPreparedStatement *)malloc(sizeof(PgPreparedStatement) + query_len + num_parameters);
+	ps->query_len = query_len;
+	ps->param_len = num_parameters;
+	ps->query_id = next_unique_query_id;
+	ps->use_count = 0;
+	memcpy(ps->query_with_params, pkt->query, query_len);
+	memcpy(&ps->query_with_params[query_len], pkt->parameter_types_bytes, num_parameters);
+	return ps;
+}
+
+static PgClientPreparedStatement*
+create_client_prepared_statement(char const* name, PgPreparedStatement* ps)
+{
+	size_t name_len = strlen(name) + 1;
+	PgClientPreparedStatement *client_ps = (PgClientPreparedStatement*)malloc(sizeof(PgClientPreparedStatement) + name_len);
+	memcpy(client_ps->stmt_name, name, name_len);
+	client_ps->ps = ps;
+	ps->use_count += 1;
+	return client_ps;
+}
+
+static PgServerPreparedStatement*
+create_server_prepared_statement(PgPreparedStatement* ps)
+{
+	PgServerPreparedStatement *server_ps = (PgServerPreparedStatement*)malloc(sizeof(PgServerPreparedStatement));
+	server_ps->ps = ps;
+	server_ps->query_id = ps->query_id;
+	ps->use_count += 1;
+	return server_ps;
+}
+
+static PgPreparedStatement*
+get_prepared_statement(PgParsePacket *pkt, bool* found)
+{
+	PgPreparedStatement *ps = create_prepared_statement(pkt);
+	PgPreparedStatement *old_ps = NULL;
+	HASH_FIND(hh, prepared_statements, ps->query_with_params, ps->query_len + ps->param_len, old_ps);
+	if (old_ps != NULL)
+	{
+		free(ps);
+		ps = old_ps;
+		*found = true;
+	}
+	else
+	{
+		HASH_ADD_KEYPTR(hh, prepared_statements, ps->query_with_params, ps->query_len + ps->param_len, ps);
```

Comment:
According to uthash docs, HASH_ADD_KEYPTR only needs to be used when the key is stored outside the structure. That's not the case here.
```suggestion
		HASH_ADD(hh, prepared_statements, query_with_params, ps->query_len + ps->param_len, ps);
```